### ISSUE TITLE
Unable to override `quarkus.http.port` with a `config/application.properties` file

### ISSUE THREAD

**Issue by ppatierno** [quarkusio/quarkus#31151]
Hello,
following the doc here https://quarkus.io/guides/config-reference#application-properties-file I would like to provide an application.properties file to override `quarkus.http.port` property.
I have tried different ways:

* created config/application.properties in the target Maven folder and run application with mvn quarkus:dev
* tried the above but with config/application.properties under target/classes
* create config/application.properties in target/quarkus-app and started my application with java -jar

All of them with no luck. The configuration is ignored.
What's the mistake I am making? Thanks!

**Comment by ppatierno** [comment]
I found that if I have a `config/application.properties` in the root of the project, the configuration is picked up whatever is the way to start it.
While it could make sense for quarkus:dev I didn't expect that when running the quarkus-run.jar ... I am confused

**Comment by ppatierno** [comment]
So I found that this is the reason why it's behaving this way:

https://github.com/quarkusio/quarkus/blob/main/core/runtime/src/main/java/io/quarkus/runtime/configuration/ApplicationPropertiesConfigSourceLoader.java#L43

The `user.dir` system property is used and in both cases I am running my application from the current project directory.
Anyway I was wondering that the following note in the official documentation is wrong, or?

> The config/application.properties is also available in dev mode. The file needs to be placed inside the build tool output directory (target for Maven and build/classes/java/main for Gradle)

**Comment by geoand** [comment]
`config/application.properties` is relattive to the working directory of the launched process.

@radcortez do you think we need to do anything for this one?

**Comment by ppatierno** [comment]
@geoand yeah, that's the usage of `user.dir` in the code. I think it's fine but the documentation should be addressed (I can open a PR if you want).
Regarding my other question, do you have an answer for that? Thanks!

> can I assume -Dquarkus.config.locations is going to be supported? I will use it to specify the path of the application.properties file.

**Comment by geoand** [comment]
> I think it's fine but the documentation should be addressed

I'll leave that up to @radcortez. 

> can I assume -Dquarkus.config.locations is going to be supported

I thought it already was... @radcortez ?

**Comment by ppatierno** [comment]
> I thought it already was... @radcortez ?

no sorry, maybe I asked the wrong way :-) It's already supported, I meant ... is it something which is going to be deprecated, is it the recommended way to specify a config file, is there another better way?

**Comment by geoand** [comment]
No problem :)

As far as I know it's the recommeded way and I have not heard anything about it being deprecated

**Comment by radcortez** [comment]
> `config/application.properties` is relattive to the working directory of the launched process.
> 
> @radcortez do you think we need to do anything for this one?

I don't think so. It always worked this way.

> @geoand yeah, that's the usage of `user.dir` in the code. I think it's fine but the documentation should be addressed (I can open a PR if you want).

Feel free to open a PR if you think you can clarify it further :)

> no sorry, maybe I asked the wrong way :-) It's already supported, I meant ... is it something which is going to be deprecated, is it the recommended way to specify a config file, is there another better way?

No, it is not going to be deprecated. We learned that developers are super picky regarding config :) We provided `quarkus.config.locations` to allow reading different config files, from either relative or absolute paths. In other words, you can use any path or file name.

**Comment by ppatierno** [comment]
@radcortez actually I have found another issue ...

Imagine I have a `config/application.properties` file with `quarkus.http.port=9090` .
Running just:

```shell
mvn quarkus:dev
```

I am able to override the HTTP port from the default 8080 to 9090.

Running:
```shell
mvn quarkus:dev -Dquarkus.http.port=9999
```

I am able to override the HTTP port from the 9090 in the properties file to 9999 by the system property.
But if I specify the `quarkus.config.locations` this way:

```shell
mvn quarkus:dev -Dquarkus.config.locations=config/application.properties -Dquarkus.http.port=9999 -Dsuspend
```

(where the path is accidentally the same but could be any other one)
The system property doesn't override what I have in the properties file and I still see HTTP port 9090 being used.
Does specifying `quarkus.config.locations` have precedence on system property? I think it should not matter, or? It's about loading the application.properties file from a path but it should come after the system properties and env vars I guess.

**Comment by radcortez** [comment]
This is expected. Configuration loaded by `quarkus.config.locations` uses the ordinal value from the source used to set the property. If `quarkus.config.locations` is set on System properties (ordinal `400`), all sources loaded by `quarkus.config.locations` will also have the ordinal `400`.

If we would use the default `100` to load the sources then it wouldn't be enough to override configuration from the main `application.properties`, which users `250`. The general expectation is that `quarkus.config.locations` sources could override the main `appplication.properties` file, so we came out with the described behavior. 

You could add a `config_ordinal` property to override the source ordinal.

**Comment by ppatierno** [comment]
> This is expected. Configuration loaded by quarkus.config.locations uses the ordinal value from the source used to set the property. If quarkus.config.locations is set on System properties (ordinal 400), all sources loaded by quarkus.config.locations will also have the ordinal 400.

You are saying that its ordinal is 400 because I am using `-Dquarkus.config.locations=config/application.properties`.
If I used `QUARKUS_CONFIG_LOCATIONS= ...` via env var it would be 300?
Imho this behaviour is really confused and is it documented somewhere?

> If we would use the default 100 to load the sources then it wouldn't be enough to override configuration from the main application.properties, which users 250. The general expectation is that quarkus.config.locations sources could override the main appplication.properties file, so we came out with the described behavior.

What I am looking for is a way to specify an application.properties in a different path than just config/application.properties but leaving all the rest as it is, i.e. overriding fields via System properties, but it doesn't seem to be the case. Why don't allow to specify a different path without having the above "complicated" behaviour?

To provide you more context, as one of the Strimzi maintainers, we are migrating our Strimzi HTTP bridge to use Quarkus.
Right now the application.properties configuration file is provided via a command line `--config-file` option and I have some code working this was by retrieving the configuration from the file from a bean but not using the "Quarkus way" (ConfigMapping, ConfigProperty and so on).
In order to reach a more "Quarkus way" configuration, the first problem I was trying to address was about removing this command line option using something more Quarkus-ish so the usage of `quarkus.config.locations`.

The other problem we are facing is about properties having names using `.` as separator.
For example, the bridge configuration accepts something like `kafka.consumer.auto.offset.reset=earliest` or in general something like `kafka.consumer.<Kafka consumer property>` where there is this `kafka.consumer.` prefix + <Kafka consumer property> which is any property coming from official [Kafka consumer configuration ](https://kafka.apache.org/documentation/#consumerconfigs) where `.` is used as separator.
I know it's not possible in Quarkus (unless you use double quotes around the name, but it's not feasible to ask users to change their configuration file because of this). I also opened this discussion in Quarkus [here](https://github.com/quarkusio/quarkus/discussions/31156).
I explored different potential solutions but no one matches what we need.
Do you have anything in mind for this "migration"?

**Comment by radcortez** [comment]
> You are saying that its ordinal is 400 because I am using `-Dquarkus.config.locations=config/application.properties`.
> If I used `QUARKUS_CONFIG_LOCATIONS= ...` via env var it would be 300?

Correct!

> Imho this behaviour is really confused and is it documented somewhere?

Yes, it is documented in SmallRye Config here: https://smallrye.io/smallrye-config/3.1.2/config-sources/locations/. Yes, I understand is a bit hidden. We always struggle with how to expose documentation of libraries into Quarkus without having to duplicate everything.

Initially, it didn't work this way. The additional sources were loaded with the default ordinal of `100`, which was insufficient to override the main `application.properties`, which sits at `250`. This caused some frustration to users. We thought it was very arbitrary to assign a specific ordinal, so we settled with using the ordinal from the source that provided `quarkus.config.locations`. Do you have a better idea? I'm happy to revisit this and improve it.

> What I am looking for is a way to specify an application.properties in a different path than just config/application.properties but leaving all the rest as it is, i.e. overriding fields via System properties, but it doesn't seem to be the case. Why don't allow to specify a different path without having the above "complicated" behaviour?

You can still do it. You can override this behavior by adding a `config_ordinal=299` (or something between 251 and 299) to the external ´application.properties´ file. This will ignore the ordinal coming from System Properties and use the one set in `config_ordinal`. Config in this file will override the main `application.properties`, but will not get overridden by ENV vars or System Properties.

> Do you have anything in mind for this "migration"?

Maybe an interceptor can do the trick?
https://smallrye.io/smallrye-config/3.1.2/extensions/interceptors/

If you know that everything after `kafka.consumer` is the key, including dots, then you can use an interceptor to rewrite the key to add the quotes so it gets correctly mapped. We use these interceptors for profiles, relocations, and fallbacks. It should also work for this case.

**Comment by ppatierno** [comment]
> ....  but will not get overridden by ENV vars or System Properties.

Actually I was able to override via a System Properties, which is what I wanted. I mean, setting `config_ordinal=299` in the file and then running:

```shell
mvn quarkus:dev -Dquarkus.config.locations=config/application.properties -Dquarkus.http.port=9999
```

I can see HTTP port 9999 being used (instead 9090 in the file), so it's working as I expected but you were saying the opposite above?
Said that, why not using an ordinal > 251 and < 299 (not 100, not the one from env var or sys properties) by default?
I think it's a good compromise with overriding the main `application.properties` but being overridden by env vars and sys props. Maybe there is a good reason I cannot see why you didn't do that way?

> Maybe an interceptor can do the trick?

I will take a look thanks!

**Comment by radcortez** [comment]
> Actually I was able to override via a System Properties, which is what I wanted. I mean, setting `config_ordinal=299` in the file and then running:

Yes, this works, and it was my recommendation :)

>  Said that, why not using an ordinal > 251 and < 299 (not 100, not the one from env var or sys properties) by default? I think it's a good compromise with overriding the main `application.properties` but being overridden by env vars and sys props. Maybe there is a good reason I cannot see why you didn't do that way?

Because there are other custom sources with different ordinals (and users can even create their own), it would be almost impossible to come up with a reasonable default.

**Comment by ppatierno** [comment]
> Yes, this works, and it was my recommendation :)

Yes but you said "but will not get overridden by ENV vars or System Properties." which was not true, it's working. Maybe I misled what you meant with that sentence.

> Because there are other custom sources with different ordinals (and users can even create their own), it would be almost impossible to come up with a reasonable default.

That's absolutely true. We don't know about the users ones. What are the other custom sources? 
Anyway from my pov, the goal is being able to:

- specify a different path for a config file
- override the main application.properties
- being overridden by env vars and system properties

All the above let me think that a good ordinal is between 261 and 299, unless the "custom sources" you mentioned are something I am missing (which I don't know what they are). Notice that I mentioned 261 and not 251 referring to the table here https://quarkus.io/guides/config-extending-support with ordinals, where I don't see any other sources in between.

Regarding interceptor I had it working.
I am able to add the quotes and having the Map with the values.
I am also able to do a trick like this: we already have an `http.port` configuration in our properties file, I can intercept that and returning a `ConfigValue` with the `quarkus.http.port` as key instead but with our value :-) This keeps backward compatibility with what users set for the HTTP port.
Again the only thing I don't like is setting a config_ordinal=299 in the properties file. Is there no way to make this change somehow programmatically and hiding it to the user?

**Comment by radcortez** [comment]
> Yes but you said "but will not get overridden by ENV vars or System Properties." which was not true, it's working. Maybe I misled what you meant with that sentence.

Ah what I meant was that the file in the `config_ordinal=[251-299]` would not override the env or system properties, due to the order being lower of course. Sorry for the mislead.

> That's absolutely true. We don't know about the users ones. What are the other custom sources?

For instance, YAML (with `application.yaml`) uses `255` and `265`, Consul uses `270` and JDBC `400`.

> All the above let me think that a good ordinal is between 261 and 299, unless the "custom sources" you mentioned are something I am missing (which I don't know what they are). Notice that I mentioned 261 and not 251 referring to the table here https://quarkus.io/guides/config-extending-support with ordinals, where I don't see any other sources in between.

We have this project with additional sources: 
https://github.com/quarkiverse/quarkus-config-extensions

SmallRye Config also provides some additional sources:
https://smallrye.io/smallrye-config/3.1.2/config-sources/custom/

> Regarding interceptor I had it working. I am able to add the quotes and having the Map with the values. I am also able to do a trick like this: we already have an `http.port` configuration in our properties file, I can intercept that and returning a `ConfigValue` with the `quarkus.http.port` as key instead but with our value :-) This keeps backward compatibility with what users set for the HTTP port.

Yes, the interceptors are very powerful in that regard and we implement all the config resolutions on top of them. For relocate and fallback we already support them out of the box:

https://smallrye.io/smallrye-config/3.1.2/extensions/fallback/
https://smallrye.io/smallrye-config/3.1.2/extensions/relocate/

I recommend using ours because they already account for things like profile resolution  and they can even take functions to transform the keys :)

Again the only thing I don't like is setting a config_ordinal=299 in the properties file. Is there no way to make this change somehow programmatically and hiding it to the user?

Not currently. The only way would be to provide your own way to load additional files and drop the rule that takes the ordinal. The code is super generic, so you only need to implement one class. Check here:

https://github.com/smallrye/smallrye-config/blob/56b9c972cd471d1162ad6726dacb12ffb7946b1d/implementation/src/main/java/io/smallrye/config/AbstractLocationConfigSourceFactory.java#L17
https://github.com/smallrye/smallrye-config/blob/a10d000651633db090e1d3a2d057ebab1e8d8004/implementation/src/main/java/io/smallrye/config/PropertiesLocationConfigSourceFactory.java#L8

**Comment by radcortez** [comment]
I don't think there's anything to do here.

**Comment by Tazzmanian** [comment]
For any new user to quarkus such as myself, I had the same issue, I couldn't override the port (resource/application.properties -> quarkus.http.port). When running with the CLI I was just getting that 8080 is in use, same with mvn quarkus:dev, but when I tried `mvn quarkus:dev -D"quarkus.http.port=8083"` I got that [error]: Build step ..., use @jakarta.inject.Inject instead. After changing it the port override works. (FYI: I am using VS Code without any extensions for study purpose)
